Reader: Auto-update filter chips when subscriptions are updated

dvdchr opened the pull request:
Refs #22601

This adds an observer class, `ReaderTopicChangeObserver,` to the `FilterProvider`, which will notify the provider whenever Core Data updates the related model (such as `ReaderTagTopic` or `ReaderSiteTopic`). However, with this, I'd need to differentiate the data fetching method (i.e., locally or remotely) because fetching the data remotely will cause the observer to react and, in turn, trigger the fetching process again—causing an infinite loop!

It's not ideal since the `Provider` implementation needs to respect the `localOnly` parameter, which makes it a bit prone to error. Still, I'll note this down for future improvement since we will need to keep `FilterProvider` as context-free as possible. 

In addition, through https://github.com/wordpress-mobile/WordPress-iOS/commit/0887244a717648a3269d13ddaf612b47b8c33ccd, this also resolves a bug(?) where the tag's `following` property is not set to `true` after subscribing to it from the Manage flow. This little issue made it look like the observer was not working properly... 😅 

## To test

### Case 1: Manage flow

- Launch the Jetpack app and move to the Reader tab.
- Switch to the Subscriptions stream.
- Tap the `Tags` filter chip, and then Edit.
- Add a new tag, and tap Done.
- 🔎 Verify that the new tag appears on the filter sheet.
- 🔎 Verify that the number on the tag filter chip is updated.
- Tap Edit again.
- Remove the tag that was added previously, and tap Done.
- 🔎 Verify that the tag is removed from the filter sheet.
- 🔎 Verify that the number on the tag filter chip is updated.
- 🔁 Repeat the steps above for the `Blogs` filter chip.

### Case 2: Search flow

- Launch the Jetpack app and move to the Reader tab.
- Switch to the Subscriptions stream.
- Tap on the Search icon, and search for a site (e.g., "Matt Mullenweg")
- Select the site on the `Blogs` tab, and tap Subscribe.
- Return to the Reader stream.
- 🔎 Verify that the number on the filter chip is updated.

### Case 3: Universal Link

- Launch the Jetpack app and move to the Reader tab.
- Switch to the Subscriptions stream.
- Tap a universal link to a post detail from an unsubscribed blog.
  - 💡 **Tip:** If you're on a simulator, from the terminal you can simulate a deeplink by entering:
    ```
    xcrun simctl openurl booted '<universal-link-post-here>'
    ```
- Tap `Subscribe` on the top right of Post Detail.
- Tap back to return to the Reader stream.
- 🔎 Verify that the number on the filter chip is updated.

## Regression Notes
1. Potential unintended areas of impact
The change is isolated to the `FilterProvider` which is only used in the Reader navigation bar, but may have risk since it now subscribes to Core Data changes.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manually tested the changes.

3. What automated tests I added (or what prevented me from doing so)
N/A.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [x] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

dvdchr committed:
Add an observer class for ReaderTopic

dvdchr committed:
Locally update the tag's `following` property after adding a tag

Previously, adding a tag from the Manage flow does not update the
tag's `following` property locally until the next time it's fetched.

Additionally, the model's `following` property seems to be affecting
the value of `showInMenu`. I've also added a convenience method that
ensures that `showInMenu` is also updated when `following` is updated.

dvdchr committed:
Update FilterProvider to react to topic changes

dvdchr committed:
Remove the completion block usage from ReaderManagePresenter

dvdchr committed:
Add docs for the Provider typealias

dvdchr committed:
Update unit tests

dvdchr committed:
Add more docs

dvdchr committed:
Add docs for the toggleFollowing convenience method

wargcm reviewed (approved):
### Case 1: Manage flow

 ✅ Following/unfollowing tags from the manage screen updated correctly.

⚠️ I did find that following tags from `Discover more tags` did not update the chip/list until tapping on the chip again:

https://github.com/wordpress-mobile/WordPress-iOS/assets/2454408/ab774797-495e-474a-b36d-d4a66a2102c9

⚠️ Removing a followed blog works from the manage screen, but when I added the blog back, the list and number did not update until tapping on the blog chip again. I recorded a video of this, but I'll share it over Slack due to private P2s showing.

### Case 2: Search flow

✅ Worked as described.

### Case 3: Universal Link

⚠️ No update until tapping on the blog chip:

https://github.com/wordpress-mobile/WordPress-iOS/assets/2454408/f7e016f3-eaa5-4de1-9521-c7f98ec18427

---

Even though I had some issues with it, I'm still going to approve because it's in a better state than it was previously. I can re-review this PR if you make more changes to it, or we can possibly fix these in a future PR.

dvdchr commented:
> - ⚠️ I did find that following tags from Discover more tags did not update the chip/list until tapping on the chip again:
> - ⚠️ Removing a followed blog works from the manage screen, but when I added the blog back, the list and number did not update until tapping on the blog chip again
> - ⚠️ No update until tapping on the blog chip

Good catch! Since this is related to Core Data, I'd prefer pushing this to the next release instead to avoid unwanted risks. 

It seems that relying on the topic's `following` property isn't enough. My assumption was that we were doing optimistic Core Data updates to the `following` property, but like `ReaderTopicService.followTagNamed:`, it seems like they only make a `POST` request and rely on the next sync / pull-to-refresh to update topics' state.

dvdchr committed:
Trigger observer delegate call on topic refresh

Previously, the observer delegate is only called when there are objects
of type `Topic` being updated (`NSUpdatedObjectsKey`). However, it turns
out that, in some cases, are not up-to-date enough.

To ensure that the UI updates after the models' changes are persisted,
the observer now also watches for `NSRefreshedObjectsKey`.

dvdchr commented:
Hi @wargcm 👋 , it turns out that listening `NSUpdatedObjectsKey` is not enough; in some cases (like adding tags from `Discover more tags`), it doesn't guarantee that the models have the latest changes when updating the UI. To solve this, I added `NSRefreshedObjectsKey` to ensure that the objects are in their most recent state.

This is now ready for your second round of review!

wargcm reviewed (approved):
These changes address the issues I had previously. Nice work, LGTM! :shipit:
